#107 Setup Danger

kevnm67 opened the pull request:
## Pull request checklist

- [x] All tests pass. 
- [x] Demo project builds and runs.
- [x] I have resolved merge conflicts.
- [x] I have followed the [coding style](https://github.com/jessesquires/HowToContribute#style-guidelines) and [contributing guidelines](https://github.com/jessesquires/HowToContribute).

#### This fixes issue #
#107 Setup Danger

## What's in this pull request?
- [Added] Dangerfile with related files for installing required plugins.
- [Added] Codeowners file.
- [Updated] Travis config to setup danger and exec danger at the end of the script phase.

kevnm67 committed:
Added Dangerfile and updated travis config

- [Added] Dangerfile with related files for installing required plugins.
- [Added] Codeowners file.
- [Updated] Travis config to setup danger and exec danger at the end of the script phase.

kevnm67 committed:
Fixed build error for test target

kevnm67 committed:
Fixed range exception crash

- [Added] script to test danger locally.

kevnm67 commented:
@jessesquires I have danger working locally and in the [PR I created from my forked version](https://github.com/kevnm67/JSQDataSourcesKit/pull/1#issuecomment-470643019).  A few things to note:

1.  The test suite had compiler errors.  It looks like it's been happening for a few months according to travis.  I fixed those in this branch.  Not sure if you want these broken out to a separate PR? or...

2.  After fixing the above I ran into 2 more issues: A failed test case and a range exception.  I fixed the range exception and will fix the failed test case (if you want).

Im going to assume you'd want those in a separate PR unless you say otherwise.

jessesquires commented:
Hey @kevnm67 -- this is looking great! 💯 🎉 

Commented on #121. Sorry about that! That was from a recent commit. All issues on `develop` should be resolved now. 

I've also added `CODEOWNERS` on `develop` (to make this diff smaller).

Could you rebase/merge with `develop` and revert the code and swiftlint changes?

jessesquires commented on .ruby-version:
```diff
@@ -0,0 +1 @@
+2.5.1
```
do we need this?

i think in `.travis.yml` we can just specify `rvm: 2.5.1`

(would rather keep this in one place)

jessesquires replied:
ah, looks like we don't need this:

https://docs.travis-ci.com/user/languages/ruby/#using-ruby-version

kevnm67 replied:
You can use either.  My reasoning: Ensure the ruby version is locked/consistent whether running on travis or locally.  

Which do you prefer?

jessesquires replied:
let's go with specifying in only the `.travis.yml` file for now. if needed, we can revisit.

jessesquires commented on .travis.yml:
```diff
@@ -1,5 +1,12 @@
-language: objective-c
+language: swift
```
hm. does this work? travis-ci docs still say to use `objective-c` 😄

kevnm67 replied:
Actually, I don't think we need to specify the language at all. This works:  
```yaml
os: osx
osx_image: xcode10.1
```

kevnm67 replied:
> hm. does this work? travis-ci docs still say to use `objective-c` 😄

Magic :) I can't remember where/when I read about this setting but have used it for a year or so.  See my previous comment, though.

jessesquires commented on .travis.yml:
```diff
@@ -1,5 +1,12 @@
+  homebrew:
+    packages:
+    - swiftlint
+    update: true
```
wow, this is new. 💯 TIL

jessesquires commented on .travis.yml:
```diff
@@ -22,9 +29,8 @@ env:
-   - gem install cocoapods --pre --no-rdoc --no-ri --no-document --quiet
-   - brew update
-   - brew outdated swiftlint || brew upgrade swiftlint
+  - rvm install 2.5.1
```
per comment above, we can remove this

jessesquires commented on .travis.yml:
```diff
@@ -1,5 +1,12 @@
+    - swiftlint
+    update: true
+cache:
+  bundler: true
```
would prefer using `cache: bundler`

jessesquires commented on .travis.yml:
```diff
@@ -22,9 +29,8 @@ env:
-   - brew update
-   - brew outdated swiftlint || brew upgrade swiftlint
+  - rvm install 2.5.1
+  - bundle install --without=documentation
```
we don't need this. travis will call `bundle install` automatically

let's update `before_install:` to be:

```
before_install:
    - gem update --system
    - gem install bundler
```

(to install bundler 2.0 https://docs.travis-ci.com/user/languages/ruby/#bundler-20)

jessesquires replied:
for below, L40, let's `bundle exec` the `pod lib lint` line

jessesquires commented on .travis.yml:
```diff
@@ -48,6 +54,8 @@ script:
       xcodebuild test -project JSQDataSourcesKit.xcodeproj -scheme JSQDataSourcesKit -destination "platform=iOS Simulator,name=iPhone X" CODE_SIGNING_REQUIRED=NO;
   fi
+  
+- bundle exec danger
```
👏

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+# end
+
+# Let people say that this isn't worth a CHANGELOG entry in the PR if they choose
+declared_trivial = (github.pr_title + github.pr_body).include?("#trivial") || !has_app_changes
```
let's remove this

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+no_changelog_entry = !git.modified_files.include?("CHANGELOG.md")
+
+if has_app_changes && no_changelog_entry && !declared_trivial
+  warn("Any changes to library code should be reflected in the Changelog. Please consider adding a note there and adhere to the [Changelog Guidelines](https://github.com/jessesquires/JSQDataSourcesKit/CHANGELOG.md).")
```
here, let's say, "There's no changelog entry. Do you need to add one?"

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+# 	auto_label.remove("WIP")
+# 	# If you want to delete label
+# 	# auto_label.delete("WIP")
+# end
```
let's remove L35 to here 😄

kevnm67 replied:
NOTE: Was commenting out bc I needed the label on this repo + the danger token.  If you want to auto-label PR's for certain keywords, let me know ;)
Removed.

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+warn("Big PR, try to keep changes smaller if you can") if git.lines_of_code > 500
+
+# Make it more obvious that a PR is a work in progress and shouldn't be merged yet
+warn("PR is classed as Work in Progress") if github.pr_title.include? "[WIP]"
```
let's remove this

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+
+warn('Changes to CI/CD files') if didModify(@DANGER_CI_CD_FILES)
+warn('Changes to dependency related files') if didModify(@DANGER_DEPENDENCY_FILES)
+warn("Big PR, try to keep changes smaller if you can") if git.lines_of_code > 500
```
let's say: "This is a large pull request! Can you break it up into multiple smaller ones instead?"

kevnm67 replied:
Done.

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+
+has_app_changes = !git.added_files.grep(/Source.*\.swift/).empty?
```
could we call this ~~`has_lib_changes`~~ `has_source_changes` ?

kevnm67 replied:
of course! Changed.

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+  prose.ignore_acronyms = true
+  prose.ignore_numbers = true
+  prose.check_spelling modified_prose_lintable_files
+end
```
nice! 👏

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+end
+
+# Run danger-prose to lint and check spelling.
+modified_prose_lintable_files = (git.added_files.grep(%r{.*\.md/}) + git.modified_files.grep(%r{.*\.md/}))
```
could we name this `markdown_files` ?

kevnm67 replied:
This was a guess as to what you'd want to lint using prose.  Is there anything else I should add? or exclude?

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,82 @@
+files_to_lint.reject! { |f| f.start_with?('scripts/', 'docs/') }
+swiftlint.lint_files(files_to_lint,
+	additional_swiftlint_args: '--no-force-exclude'
+)
```
I think we can update all the swiftlint stuff to be:

```ruby
swiftlint.verbose = true
swiftlint.config_file = './.swiftlint.yml'
swiftlint.lint_files(inline_mode: true, fail_on_error: true)
```

kevnm67 replied:
Ok, ill try that 🤞.  I've had issues with two other projects and the way I wrote it here is what's been working (although not ideal, obviously).

Out of scope, but have you used [code climate](https://codeclimate.com/quality)?  I use it in multiple projects and it works great with swiftlint + inline comments.

jessesquires commented on Gemfile:
```diff
@@ -0,0 +1,9 @@
+source "https://rubygems.org"
+
+gem 'cocoapods'
+gem 'danger'
```
let's specify versions for these:

```spotify
gem 'cocoapods', '~> 1.6.0'
gem 'danger', '~> 5.0'
```

kevnm67 replied:
Good call. Done

jessesquires commented on Gemfile:
```diff
@@ -0,0 +1,9 @@
+gem 'danger'
+gem 'danger-swiftlint'
+gem 'danger-xcode_summary'
+gem 'xcpretty-json-formatter'
```
might as well explicitly specify `xcpretty` too

jessesquires commented on scripts/test_danger.sh:
```diff
@@ -0,0 +1,23 @@
+# Run danger locally #
+######################
+
+run_against_pr 1
```
nice! 👏

kevnm67 replied:
😊

jessesquires commented on scripts/test_danger.sh:
```diff
@@ -0,0 +1,23 @@
+    cd ..; bundle install
+
+    # Use the provided PR to run danger against.
+    bundle exec danger pr https://github.com/kevnm67/JSQDataSourcesKit/pull/$(get_pr_value $1)
```
this should be `jessesquires` not `kevnm67`, right? 😄

kevnm67 replied:
:) Indeed... meant to switch this over after you added the danger token.  Ill fix this now.

jessesquires commented on scripts/test_danger.sh:
```diff
@@ -0,0 +1,23 @@
+    if [ $# = 1 ] ; then
+        echo $1
+    else
+        echo "1"
```
instead , maybe we could just error if you don't provide a PR number?

kevnm67 replied:
Sure, that makes sense.  Probably a better approach as well ;)

jessesquires reviewed (commented):
Left a bunch of inline comments but overall this is awesome! 💯 🥇 

Thanks so much @kevnm67 !

kevnm67 committed:
Updated based on jessesquires code review

kevnm67 commented:
> Hey @kevnm67 -- this is looking great! 💯 🎉
> 
> Commented on #121. Sorry about that! That was from a recent commit. All issues on `develop` should be resolved now.
> 
> I've also added `CODEOWNERS` on `develop` (to make this diff smaller).
> 
> Could you rebase/merge with `develop` and revert the code and swiftlint changes?

Thanks!
No worries.  Thanks for adding the file in develop to ↓ the diff size ;)

Done.  The only reason I updated the swiftlint config was in response to swiftlint compiler warnings.  Something about not matching... reverted.  Weird, though.  Looks like swiftlint's repo using those rules 🤷‍♂️

kevnm67 commented:
@jessesquires Looks like travis is still missing the danger token environment variable:
```For your GitHub repo, you need to expose: DANGER_GITHUB_API_TOKEN```

Also, you may want to update your repo settings for travis and migrate to Travis to GitHub Apps now that [services has been deprecated](https://developer.github.com/changes/2018-04-25-github-services-deprecation/).

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,59 @@
+
+has_source_changes = !git.added_files.grep(/Source.*\.swift/).empty?
```
this should just be `/Source/`

jessesquires commented on Dangerfile:
```diff
@@ -0,0 +1,59 @@
+### WARNINGS
+
+warn('Changes to CI/CD files') if didModify(@DANGER_CI_CD_FILES)
+warn('Changes to dependency related files') if didModify(@DANGER_DEPENDENCY_FILES)
```
Thinking about this more...

Actually, I don't think we need these specific warnings. These are rarely modified by contributors anyway, so I should always notice. If it ever becomes an issue, we can revisit.

jessesquires commented:
@kevnm67 Thanks for making those changes! Left a couple more comments.

>  Looks like travis is still missing the danger token environment variable

Double-checked. It's definitely there. I think we might need to merge this first for it to register. After these last changes, let's get this merged, then we can do another PR for #122 (and maybe some other random PRs to test).

> Also, you may want to update your repo settings for travis and migrate to Travis to GitHub Apps now that services has been deprecated.

Ah! 💯 I'll look into this!

kevnm67 commented:
@jessesquires Made all the changes from your latest review ;)

kevnm67 committed:
Merge branch 'develop' into feature/issue_107_setup-danger

jessesquires commented:
Thanks @kevnm67 !

Just sent you an invite to be a collaborator on this repo 😄

kevnm67 commented:
Awesome!  Thanks @jessesquires, I appreciate it.
